### PR TITLE
Darwin chip tool cli clusters

### LINKED ISSUE #14798: Add README to chip-tool-darwin
#### Problem
Add a README to chip-tool-darwin

### PR DESCRIPTION
**Pull request by krypton36**
#### Problem
chip-tool-darwin currently does not have cluster commands and missing README.

#### Change overview
- Add Zap file to generate cluster commands using CHIP.framework
- Add Zap file to generate README
- Generated README and Cluster Commands.

#### Testing
- Paired all-cluster-app
- Ruan several commands.
- Ran several  reads
- Ran several writes

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
Fast tracking adding a new tool.

File: .github/.wordlist.txt
Comment (andy31415) [inline .github/.wordlist.txt:944]: This does look like a typo

**Review by vivien-apple — COMMENTED** [review]
I did a PR to your branch with some of the changes mentioned in this review: https://github.com/krypton36/connectedhomeip/pull/33

File: examples/chip-tool-darwin/templates/README.zapt
Comment (andy31415) [inline examples/chip-tool-darwin/templates/README.zapt:125]: Could we avoid having a ZAPT for readme files even?

This is too much generated code.

**Review by andy31415 — DISMISSED** [review]
I think having README.md generated via ZAP is too much.

Our path should be generating files at compile time. This goes against that: more likely chances to have merge conflicts and generates large outputs in our source tree.

File: examples/chip-tool-darwin/templates/commands.zapt
Comment (andy31415) [inline examples/chip-tool-darwin/templates/commands.zapt:285]: How about naming things 'subscription' and similar to be more in line with matter spec?

File: examples/chip-tool-darwin/commands/clusters/ModelCommandBridge.mm
Comment (carol-apple) [inline examples/chip-tool-darwin/commands/clusters/ModelCommandBridge.mm:31]:
```suggestion
    ChipLogProgress(chipTool, "Sending command to node 0x" ChipLogFormatX64, ChipLogValueX64(mNodeId));
```

File: examples/chip-tool-darwin/README.md
Comment (vivien-apple) [reply examples/chip-tool-darwin/README.md:5]: It also seems that `signgin` has been added to the words list !

File: examples/chip-tool-darwin/templates/README.zapt
Comment (krypton36) [reply examples/chip-tool-darwin/templates/README.zapt]: Done

File: examples/chip-tool-darwin/templates/README.zapt
Comment (krypton36) [reply examples/chip-tool-darwin/templates/README.zapt]: Ok.

File: examples/chip-tool-darwin/commands/clusters/ModelCommandBridge.mm
Comment (krypton36) [reply examples/chip-tool-darwin/commands/clusters/ModelCommandBridge.mm]: Done. I removed anything that was not implemented.

**Review by woody-apple — APPROVED** [review]
Fast tracking platform changes by a platform maintainer.

**Comment by woody-apple** [comment]
> I think having README.md generated via ZAP is too much.
> 
> Our path should be generating files at compile time. This goes against that: more likely chances to have merge conflicts and generates large outputs in our source tree.

@andy31415 This is just following current practices re: ZAP - which do include a zap generation phase. I'm not sure how this increases any chances of merge conflicts, given this piggy backs on what exists. 

Suggest this is allowed to proceed to unblock testing.

**Comment by andy31415** [comment]
> > I think having README.md generated via ZAP is too much.
> > Our path should be generating files at compile time. This goes against that: more likely chances to have merge conflicts and generates large outputs in our source tree.
> 
> @andy31415 This is just following current practices re: ZAP - which do include a zap generation phase. I'm not sure how this increases any chances of merge conflicts, given this piggy backs on what exists.
> 
> Suggest this is allowed to proceed to unblock testing.

I strongly believe README.md should not be auto-generated.

**Comment by woody-apple** [comment]
> I strongly believe README.md should not be auto-generated. I do not believe that the value of having the command list in the readme is worth the cost in maintenance and complexity.

Given how out of date our documentation is, this seems like a nice solve in the interim, using the tools we have available. I also am not sure how this will cause more conflicts. Can you please explain? This is all a single downstream item that gets regenerated, not like our others that have cross dependencies.

File: examples/chip-tool-darwin/templates/README.zapt
Comment (krypton36) [reply examples/chip-tool-darwin/templates/README.zapt:3]: I took Vivien's advice and placed the generated files to avoid any code gen conflicts.

**Comment by krypton36** [comment]
> > > I think having README.md generated via ZAP is too much.
> > > Our path should be generating files at compile time. This goes against that: more likely chances to have merge conflicts and generates large outputs in our source tree.
> > 
> > 
> > @andy31415 This is just following current practices re: ZAP - which do include a zap generation phase. I'm not sure how this increases any chances of merge conflicts, given this piggy backs on what exists.
> > Suggest this is allowed to proceed to unblock testing.
> 
> I strongly believe README.md should not be auto-generated. I do not believe that the value of having the command list in the readme is worth the cost in maintenance and complexity.

I took @vivien-apple advice and put the generated portion into the zzz_generated folder to avoid and gen conflicts.

File: examples/chip-tool-darwin/commands/clusters/ModelCommandBridge.mm
Comment (bzbarsky-apple) [reply examples/chip-tool-darwin/commands/clusters/ModelCommandBridge.mm]: I think the right thing to do is probably:
1. Use ChipLogError.
2. Have a mode (controlled by what, not sure) where ChipLogError gets logged via NSLog.

Comment (krypton36) [reply examples/chip-tool-darwin/commands/clusters/ModelCommandBridge.mm]: Created an [issue](https://github.com/project-chip/connectedhomeip/issues/14909) for future implementation.

File: examples/chip-tool-darwin/templates/README.zapt
Comment (andy31415) [reply examples/chip-tool-darwin/templates/README.zapt:3]: Please do not generate README.md 
The readme.md can describe how to use the tool and the tool can output list commands or autocomplete.

An autogenerated readme.md cannot really give user-friendly commands, it only is a flat list of available commands, so I am claiming the benefit we get from this is minimal as it is just a list that is not a friendly to humans as there is no detail on what commands do or what they are used for.

**Comment by andy31415** [comment]
> > > > I think having README.md generated via ZAP is too much.
> > > > Our path should be generating files at compile time. This goes against that: more likely chances to have merge conflicts and generates large outputs in our source tree.
> > > 
> > > 
> > > @andy31415 This is just following current practices re: ZAP - which do include a zap generation phase. I'm not sure how this increases any chances of merge conflicts, given this piggy backs on what exists.
> > > Suggest this is allowed to proceed to unblock testing.
> > 
> > 
> > I strongly believe README.md should not be auto-generated. I do not believe that the value of having the command list in the readme is worth the cost in maintenance and complexity.
> 
> I took @vivien-apple advice and put the generated portion into the zzz_generated folder to avoid and gen conflicts.

Please remove the README.md generation out of the PR completely (regardless of location), then it should be good to go.

**Comment by woody-apple** [comment]
> > I took @vivien-apple advice and put the generated portion into the zzz_generated folder to avoid and gen conflicts.
> 
> Please remove the README.md generation out of the PR completely (regardless of location), then it should be good to go.

I'm really not sure what policy disallows adding this? Why are we blocking changes here adding up to date documentation?

**Comment by andy31415** [comment]
> > > I took @vivien-apple advice and put the generated portion into the zzz_generated folder to avoid and gen conflicts.
> > 
> > 
> > Please remove the README.md generation out of the PR completely (regardless of location), then it should be good to go.
> 
> I'm really not sure what policy disallows adding this? Why are we blocking changes here adding up to date documentation?
> 
> Edit to add:
> 
> What policy prohibits using zap and code generation?
> 
> I understand certain frustrations here, however, this seems to be a relatively straight forward change, following normal/existing code generation for platforms and tools, and unless there's precedent here, we should let this through.

There is no policy that prohibts zap code generation usage of zap. 

As a code reviewer, it is my technical opinion that auto generating README/text files into the code provides little benefit, hence my 'changes requested'. As an analogy taken to the extreme, there is no policy of "do not write C++ code" (that would be crazy), yet code reviews can reject C++ code if they feel its use is inappropriate.

**Comment by andy31415** [comment]
I would even be ok if you make the output of `--help` or similar output available commands. I was assuming we already have that. If we do not, lets add it. Whenever someone builds or tries a tool, `--help` or `-h` or `-?` is the first thing they run.

Please don't place README.md into various places and make them autogen. I am against that approach.

**Comment by andy31415** [comment]
As an aside, what is the difference beteween `chip-tool` and `chip-tool-darwin`? Both seem CLI. Would chip-tool not work on darwin as well? Why do both exist?

**Comment by woody-apple** [comment]
> As an aside, what is the difference beteween `chip-tool` and `chip-tool-darwin`? Both seem CLI. Would chip-tool not work on darwin as well? Why do both exist?

I have the same concern, however - given maybe we can file an issue here to help things move forward?

**Comment by woody-apple** [comment]
> > As an aside, what is the difference beteween `chip-tool` and `chip-tool-darwin`? Both seem CLI. Would chip-tool not work on darwin as well? Why do both exist?
> 
> I have the same concern, however - given maybe we can file an issue here to help things move forward?

Filed: https://github.com/project-chip/connectedhomeip/issues/14987

**Comment by vivien-apple** [comment]
> As an aside, what is the difference beteween `chip-tool` and `chip-tool-darwin`? Both seem CLI. Would chip-tool not work on darwin as well? Why do both exist?

`chip-tool` uses the C++ API and as such it supports some APIs that we may not support in the near future on the Objc side.
`chip-tool-darwin` uses the Objc++ API from `src/darwin/Framework` and all the wrapping that happens on top of the C++ layer.

To give you a more concrete example of some difference between what the C++ layer allows that is not supported on the Objc++ side. Using the `Any` cluster of `chip-tool` let you build IM packets from JSON specified on the command line, notably for command arguments of attributes content when writing.

That said, `chip-tool-darwin` will also run the clusters tests that currently belongs to `src/darwin/Framework` under `./scripts/tests/run_test_suite.py` (see https://github.com/vivien-apple/connectedhomeip-1/tree/ChipToolDarwin_GetChipToolDarwinToWorksWithTheTestSuiteRunner/src/app)

That would allow to get the test runs closer by using the same environment. If you look at the CI today, you will notice a lot of the "Build Darwin" tasks are red. It happens because #14606 has introduced some new `Subscribe` command to the test suites. 

I suspect that it happens randomly because while those subscriptions are closed on the client side, the server waits for some timeouts to close them on its own side (that's just how this stuff works not a bug).
But now, why does it happens only for "Build Darwin" CI and not the other CI tests ? Because "Darwin CI" is running a single instance of the "all-clusters-app" for the whole set of tests - so the test environment is different and that's just more work...

And maybe I'm completely wrong, and that's not the root cause. But the fact that the test environment  *could* be the cause is an issue.

**Comment by vivien-apple** [comment]
| For the command lists, I believe chip-tool is self-documenting and can auto-complete. I would like documentation to describe how to use that and not try to duplicate things.

I would argue that `chip-tool` would even be more self-documenting is we do use the information from the XML that (in theory) describes the cluster and what it does, the commands and their arguments, etc...

That would also be a nice way to see when the XML lack some documentation just by using the command line tool....

File: examples/chip-tool-darwin/templates/commands.zapt
Comment (bzbarsky-apple) [inline examples/chip-tool-darwin/templates/commands.zapt:120]: Floats don't have a callback name of "Unsupported".  The only thing that does is structs.

File: examples/chip-tool-darwin/templates/commands.zapt
Comment (bzbarsky-apple) [inline examples/chip-tool-darwin/templates/commands.zapt:194]: Not a CharSpan?

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/chip-tool-darwin/templates/commands.zapt
Comment (krypton36) [reply examples/chip-tool-darwin/templates/commands.zapt]: Updated. Thank you.

File: examples/chip-tool-darwin/README.md
Comment (jmeg-sfy) [inline examples/chip-tool-darwin/README.md:82]: Barrier control is not a validated cluster but a zigbee legacy we should not anymore provide non CSA cluster
it can be used as a Sandbox but that all

File: examples/chip-tool-darwin/templates/README.zapt
Comment (krypton36) [reply examples/chip-tool-darwin/templates/README.zapt:3]: Removed.

File: examples/chip-tool-darwin/README.md
Comment (krypton36) [reply examples/chip-tool-darwin/README.md]: Removed from readme.

**Comment by andy31415** [comment]
> I removed the code gen for the README as requested.

@krypton36 @woody-apple  The file seems to actually not have been removed. I would have preffered to ask for a re-review instead of dismissing as stale review and merging (or double-checking that the review comment was indeed fully addressed)

**Pull request merged**
